dist/dist_client.go

```diff
@@ -0,0 +1,500 @@
+		req.Header.Set("Content-Range", fmt.Sprintf("%d-%d", count, count+size-1))
+		resp, err := o.client.Do(req)
+		if err != nil {
+			return "", -1, errors.Wrapf(err, "Failed posting request %v", req)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != http.StatusAccepted {
+			return "", -1, fmt.Errorf("Server returned an error %d", resp.StatusCode)
+		}
+		count += size
+		session, err = resp.Location()
+		if err != nil {
+			return "", -1, errors.Wrap(err, "Failed decoding response")
+		}
+
+		session = o.url.ResolveReference(session)
+	}
+
+	ourDigest := fmt.Sprintf("%x", digester.Sum(nil))
+	d := digest.NewDigestFromEncoded(digest.SHA256, ourDigest)
+	q := session.Query()
+	q.Set("digest", d.String())
+	session.RawQuery = q.Encode()
+	req, err := http.NewRequest(http.MethodPut, session.String(), nil)
+	if err != nil {
+		return "", -1, errors.Wrap(err, "Failed opening Put request")
+	}
+	if o.authCreds != "" {
+		req.Header.Add("Authorization", "Basic "+o.authCreds)
+	}
+	req.Header.Set("Content-Type", "application/octet-stream")
+	putResp, err := o.client.Do(req)
+	if err != nil {
+		return "", -1, errors.Wrapf(err, "Failed putting request %v", req)
+	}
+	defer putResp.Body.Close()
+	if putResp.StatusCode != http.StatusCreated {
+		return "", -1, fmt.Errorf("Server returned an error %d", putResp.StatusCode)
+	}
+
+	servDigest, ok := putResp.Header["Docker-Content-Digest"]
+	if !ok || len(servDigest) != 1 {
+		return "", -1, fmt.Errorf("Server returned incomplete headers")
+	}
+
+	blobLoc, err := putResp.Location()
+	if err != nil {
+		return "", -1, errors.Wrap(err, "Failed decoding response")
+	}
+
+	blobLoc = o.url.ResolveReference(blobLoc)
+	req, err = http.NewRequest("HEAD", blobLoc.String(), nil)
+	if err != nil {
+		return "", -1, errors.Wrap(err, "Failed opening Head request")
+	}
+	if o.authCreds != "" {
+		req.Header.Add("Authorization", "Basic "+o.authCreds)
+	}
+	resp, err := o.client.Do(req)
+	if err != nil {
+		return "", -1, errors.Wrapf(err, "Failed getting new layer %v", blobLoc)
+	}
+
+	Length, ok := resp.Header["Content-Length"]
+	if !ok || len(Length) != 1 {
+		return "", -1, fmt.Errorf("Server returned incomplete headers")
+	}
+	length, err := strconv.ParseInt(Length[0], 10, 64)
+	if err != nil {
+		return "", -1, errors.Wrap(err, "Failed decoding length in response")
+	}
+
+	if servDigest[0] != d.String() {
+		return "", -1, errors.Wrapf(err, "Server calculated digest %s, not our %s", servDigest[0], ourDigest)
+	}
+
+	// TODO dist is returning the wrong thing - the hash,
+	// not the "digest", which is "sha256:hash"
+
+	return d, length, nil
+}
+
+// ociCertDir returns a path to a directory to be consumed by
+// tlsclientconfig.SetupCertificates() depending on ctx and hostPort.
+func ociCertDir(sys *types.SystemContext, hostPort string) (string, error) {
+	if sys != nil && sys.DockerCertPath != "" {
+		return sys.DockerCertPath, nil
+	}
+
+	if sys != nil && sys.DockerPerHostCertDirPath != "" {
```

Comment:
Using the `Docker*` configuration options does not quite look good, especially not without any documentation about that effect.